packages/logs/src/logs.entry.ts

```diff
@@ -1,119 +1,163 @@
-  },
-  addLoggerGlobalContext(key: string, value: ContextValue) {
-    makeStub('addLoggerGlobalContext')
-  },
-  removeLoggerGlobalContext(key: string) {
-    makeStub('removeLoggerGlobalContext')
-  },
-  setLoggerGlobalContext(context: Context) {
-    makeStub('setLoggerGlobalContext')
-  },
-  createLogger(name: string, conf?: LoggerConfiguration): Logger {
-    makeStub('createLogger')
-    return STUBBED_LOGGER as Logger
-  },
-  getLogger(name: string): Logger | undefined {
-    makeStub('getLogger')
-    return undefined
-  },
-}
+export type Status = keyof typeof StatusType
 
-export type LogsGlobal = typeof STUBBED_LOGS
+export type LogsGlobal = ReturnType<typeof makeLogsGlobal>
 
-export const datadogLogs = makeLogsGlobal(STUBBED_LOGS)
+export const datadogLogs = makeLogsGlobal((userConfiguration) => {
+  const { configuration, errorObservable, internalMonitoring } = commonInit(userConfiguration, buildEnv)
+  return {
+    configuration,
+    errorObservable,
+    internalMonitoring,
+    session: startLoggerSession(configuration, areCookiesAuthorized(configuration.cookieOptions.secure)),
+  }
+})
 
 interface BrowserWindow extends Window {
   DD_LOGS?: LogsGlobal
 }
 
 getGlobalObject<BrowserWindow>().DD_LOGS = datadogLogs
 
-export function makeLogsGlobal(stub: LogsGlobal) {
-  const global = makeGlobal(stub)
-
+export function makeLogsGlobal(
+  baseInit: (
+    configuration: LogsUserConfiguration
+  ) => {
+    session: LoggerSession
+    configuration: Configuration
+    errorObservable: ErrorObservable
+    internalMonitoring: InternalMonitoring
+  }
+) {
   let isAlreadyInitialized = false
 
-  global.init = monitor((userConfiguration: LogsUserConfiguration) => {
-    if (!checkIsNotLocalFile() || !canInitLogs(userConfiguration)) {
-      return
-    }
+  let session: LoggerSession
+  let batch: ReturnType<typeof startLoggerBatch>
+  let configuration: Configuration
 
-    if (userConfiguration.publicApiKey) {
-      userConfiguration.clientToken = userConfiguration.publicApiKey
-      console.warn('Public API Key is deprecated. Please use Client Token instead.')
-    }
-    const isCollectingError = userConfiguration.forwardErrorsToLogs !== false
-    const logsUserConfiguration = {
-      ...userConfiguration,
-      isCollectingError,
+  let globalContext: Context = {}
+  const customLoggers: { [name: string]: Logger | undefined } = {}
+
+  const sendLogBuffered = createBufferedFunction((message: LogsMessage, currentContext: Context) => {
+    if (session.isTracked()) {
+      batch.add(message, currentContext)
     }
-    const { errorObservable, configuration, internalMonitoring } = commonInit(logsUserConfiguration, buildEnv)
-    const session = startLoggerSession(configuration, areCookiesAuthorized(mustUseSecureCookie(userConfiguration)))
-    const globalApi = startLogger(errorObservable, configuration, session, internalMonitoring)
-    assign(global, globalApi)
-    isAlreadyInitialized = true
+  })
```

Comment:
This approach works but the design of this file is not so great, it is hard to follow what is going on.
It could be better to have a strategy that allow us to better separate the different responsibilities.

With the strategy suggested in the RFC, the buffer logic is separated from the after init logic.
That could allow us to move most of the initialization logic to a separate file which would reduce the complexity of this one.

wdyt?

Reply:
Do you have concrete examples of why my code is not great, and how do you think it could be improved?

Reply:
makeLogsGlobal is too complex and has too much responsibilities: 
- it creates the public API
- it handles the init logic, but take the "base" init as a dependency, which make things harder to follow and I don't see the need for it.
- due to the createBufferedFunction strategy, it needs to have access to all the components that are required once logs is initialized (batch, internal monitoring, ...) which then couple the before init logic to the after init logic.

IMO, we could make clearer the different problematic by not mixing everything.
I've tried to illustrate what I have in mind with this commit: https://github.com/DataDog/browser-sdk/commit/7174ad1e0ffbf5c30a50b8b25d1cbc6b6cb82f81

Reply:
> take the "base" init as a dependency, which make things harder to follow and I don't see the need for it
This is needed for unit test, so we can inject a session and a configuration (`maxBatchSize`).

Thank you for your commit, I'll make something like this.